### PR TITLE
SHA-3 support (+ minor fixes & perf)

### PR DESCRIPTION
**Pull request by daverodgman**
## Description

Minor fixes on top of #5820, and perf improvement (around 1.8x for SHA3-256 on M1 Pro).

Agreed with @d3zd3z that since we've both reviewed #5820, we are OK for David to single-review these updates.

## PR checklist

Please tick as appropriate and edit the reasons (e.g.: "backport: not needed because this is a new feature")

- [x] **changelog** provided
- [x] **backport** not required - new feature
- [x] **tests** provided

### CODE REVIEW

File: include/mbedtls/md.h [hunk 1]
```diff
@@ -149,19 +152,13 @@ typedef enum {
     MBEDTLS_MD_SHA384,    /**< The SHA-384 message digest. */
     MBEDTLS_MD_SHA512,    /**< The SHA-512 message digest. */
     MBEDTLS_MD_RIPEMD160, /**< The RIPEMD-160 message digest. */
+    MBEDTLS_MD_SHA3_224,    /**< The SHA3-224 message digest. */
+    MBEDTLS_MD_SHA3_256,    /**< The SHA3-256 message digest. */
+    MBEDTLS_MD_SHA3_384,    /**< The SHA3-384 message digest. */
+    MBEDTLS_MD_SHA3_512,    /**< The SHA3-512 message digest. */
 } mbedtls_md_type_t;
 
-/* Note: this should always be >= PSA_HASH_MAX_SIZE
```
Comment (mpg) [inline include/mbedtls/md.h:154]: Why did you remove this entire note? I think it is still valid (except for the last sentence) and useful.

File: include/mbedtls/md.h [hunk 2]
```diff
@@ -112,6 +112,9 @@
 #define MBEDTLS_MD_CAN_SHA512
 #define MBEDTLS_MD_SOME_LEGACY
 #endif
+#if defined(MBEDTLS_SHA3_C)
+#define MBEDTLS_MD_CAN_SHA3
```
Comment (mpg) [inline include/mbedtls/md.h:116]: I'm not sure that's quite right. PSA has distinct algorithm identifiers for the various SHA-3 variants, hence distinct `PSA_WANT` macros. Since the `MD_CAN` macros are here to provide an abstraction above `MBEDTLS_xxx_C` and `PSA_WANT` macros, they need to be as fine-grained as the most fine-grained of the two, so PSA here.

Note: since this is a public macro, this needs fixing before the next release.

Comment (mpg) [reply include/mbedtls/md.h:116]: Created https://github.com/Mbed-TLS/mbedtls/issues/7713 and added it to the next release EPIC.

File: include/mbedtls/md.h (hunk 1 above)
Comment (daverodgman) [reply include/mbedtls/md.h:154]: Sorry, I made a mistake in the merge.

Comment (mpg) [reply include/mbedtls/md.h:154]: No worries, happens!
